File: SRC/DecodingQuantities.cpp
#include <string>
#include <vector>
#include <iostream>
#include <sstream>

#include <boost/algorithm/string.hpp>

#include "StringUtils.hpp"
#include "FileUtils.hpp"
#include "Types.hpp"

#include "DecodingQuantities.hpp"

using namespace std;

DecodingQuantities::DecodingQuantities(const char *fileName) {
  cout << "Using text decoding quantities from " << fileName << endl;
  createFromGzippedText(fileName);
}

void DecodingQuantities::createFromGzippedText(const char *fileName) {
  FileUtils::AutoGzIfstream br; br.openOrExit(fileName);
  string line;
  DataType currentType = DataType::None;
  bool parsedStates = false;
  bool parsedCSFSSamples = false;
  while (getline(br, line)) {
    vector <string> splitString; istringstream iss(line); string buf; while (iss >> buf) splitString.push_back(buf);
    if (splitString.size() == 0 || (splitString.size() == 1 && splitString[0] == "")) {
      continue;
    }
    string firstToken = splitString[0]; boost::algorithm::to_lower(firstToken);
    if (firstToken == string("states")) {
      currentType = DataType::States;
      getline(br, line);
      states = StringUtils::stoi(line);
      parsedStates = true;
      cout << "Parsed states " << states << endl;
      continue;
    } else if (firstToken == string("transitiontype")) {
      currentType = DataType::TransitionType;
      getline(br, line);
      cout << "Ignored TransitionType\n";
      continue;
    } else if (firstToken == string("csfssamples")) {
      currentType = DataType::CSFSSamples;
      parsedCSFSSamples = true;
      getline(br, line);
      CSFSSamples = StringUtils::stoi(line);
      cout << "Parsed CSFSSamples " << CSFSSamples << endl;
      CSFSmap = vector < vector < vector <float> > > (CSFSSamples - 1);
      foldedCSFSmap = vector < vector < vector <float> > > (CSFSSamples - 1);
      ascertainedCSFSmap = vector < vector < vector <float> > > (CSFSSamples - 1);
      foldedAscertainedCSFSmap = vector < vector < vector <float> > > (CSFSSamples - 1);
      continue;
    } else if (firstToken == string("timevector")) {
      currentType = DataType::TimeVector;
      getline(br, line);
      vector <string> inSplitString; istringstream iss(line); string buf; while (iss >> buf) inSplitString.push_back(buf);
      timeVector = vector <float> (inSplitString.size());
      for (unsigned int i = 0; i < inSplitString.size(); i++) {
        // cout << inSplitString[i] << endl;
        timeVector[i] = StringUtils::stof(inSplitString[i]);
      }
      cout << "Parsed TimeVector " << timeVector.size() << endl;
      continue;
    } else if (firstToken == string("sizevector")) {
      currentType = DataType::SizeVector;
      getline(br, line);
      cout << "Ignored SizeVector\n";
      continue;
    } else if (firstToken == string("expectedtimes")) {
      currentType = DataType::ExpectedTimes;
      if (!parsedStates) {
        cerr << "ERROR. Parsed " << int(currentType) << " before parsing states." << endl;
        exit(1);
      }
      getline(br, line);
      vector <string> inSplitString; istringstream iss(line); string buf; while (iss >> buf) inSplitString.push_back(buf);
      expectedTimes = vector <float> (inSplitString.size());
      if (inSplitString.size() != states) {
        cerr << "ERROR. Parsed " << inSplitString.size() + " ExpectedTimes entries for " << states << " states." << endl;
        exit(1);
      }
      for (unsigned int i = 0; i < inSplitString.size(); i++) {
        // cout << inSplitString[i] << endl;
        expectedTimes[i] = StringUtils::stof(inSplitString[i]);
      }
      cout << "Parsed ExpectedTimes " << expectedTimes.size() << endl;
      continue;
    } else if (firstToken == string("discretization")) {
      currentType = DataType::Discretization;
      if (!parsedStates) {
        cerr << "ERROR. Parsed " << int(currentType) << " before parsing states." << endl;
        exit(1);
      }
      getline(br, line);
      discretization = vector <float> (states + 1);
      vector <string> inSplitString; istringstream iss(line); string buf; while (iss >> buf) inSplitString.push_back(buf);
      if (inSplitString.size() != states + 1) {
        cerr << "ERROR. Parsed " << inSplitString.size() + " Discretization entries for " << states << " states." << endl;
        exit(1);
      }
      for (unsigned int i = 0; i < inSplitString.size(); i++) {
        discretization[i] = StringUtils::stof(inSplitString[i]);
      }
      cout << "Parsed Discretization " << discretization.size() << endl;
      continue;
    } else if (firstToken == string("classicemission")) {
      currentType = DataType::ClassicEmission;
      if (!parsedStates) {
        cerr << "ERROR. Parsed " << int(currentType) << " before parsing states." << endl;
        exit(1);
      }
      classicEmissionTable = vector < vector <float> > (2);
      for (int k = 0; k < 2; k++) {
        getline(br, line);
        vector <string> inSplitString; istringstream iss(line); string buf; while (iss >> buf) inSplitString.push_back(buf);
        if (inSplitString.size() != states) {
          cerr << "ERROR. Parsed " << inSplitString.size() + " ClassicEmission entries for " << states << " states." << endl;
          exit(1);
        }
        vector <float> thisEmission = vector <float> (states);
        for (unsigned int i = 0; i < states; i++) {
          thisEmission[i] = StringUtils::stof(inSplitString[i]);
        }
        classicEmissionTable[k] = thisEmission;
      }
      cout << "Parsed ClassicEmission " << classicEmissionTable.size() << endl;
      continue;
    } else if (firstToken == string("compressedascertainedemission")) {
      currentType = DataType::CompressedAscertainedEmission;
      if (!parsedStates) {
        cerr << "ERROR. Parsed " << int(currentType) << " before parsing states." << endl;
        exit(1);
      }
      compressedEmissionTable = vector < vector <float> > (2);
      for (int k = 0; k < 2; k++) {
        getline(br, line);
        vector <string> inSplitString; istringstream iss(line); string buf; while (iss >> buf) inSplitString.push_back(buf);
        if (inSplitString.size() != states) {
          cerr << "ERROR. Parsed " << inSplitString.size() + " CompressedAscertainedEmission entries for " << states << " states." << endl;
          exit(1);
        }
        vector <float> thisEmission = vector <float> (states);
        for (unsigned int i = 0; i < states; i++) {
          thisEmission[i] = StringUtils::stof(inSplitString[i]);
        }
        compressedEmissionTable[k] = thisEmission;
      }
      cout << "Parsed CompressedAscertainedEmission " << compressedEmissionTable.size() << endl;
      continue;
    } else if (firstToken == string("csfs")) {
      currentType = DataType::CSFS;
      if (!parsedStates || !parsedCSFSSamples) {
        cerr << "ERROR. Parsed " << int(currentType) << " before parsing states and number of CSFS samples." << endl;
        exit(1);
      }
      int undistinguishedIndex = StringUtils::stoi(splitString[1]);
      vector < vector <float> > thisCSFS = vector < vector <float> > ();
      for (int k = 0; k < 3; k++) {
        getline(br, line);
        vector <string> inSplitString; istringstream iss(line); string buf; while (iss >> buf) inSplitString.push_back(buf);
        if (inSplitString.size() != states) {
          cerr << "ERROR. Parsed " << inSplitString.size() + " CSFS entries for " << states << " states: " << line << endl;
          exit(1);
        }
        vector <float> CSFSline = vector <float> (states);
        thisCSFS.push_back(CSFSline);
        for (unsigned int i = 0; i < states; i++) {
          thisCSFS[k][i] = StringUtils::stof(inSplitString[i]);
        }
      }
      CSFSmap[undistinguishedIndex] = thisCSFS;
      // cout << "Parsed CSFS " << undistinguishedIndex << " " << CSFSmap[undistinguishedIndex].size() << " " << CSFSmap[undistinguishedIndex][0].size() << endl;
      continue;
    } else if (firstToken == string("foldedcsfs")) {
      currentType = DataType::FoldedCSFS;
      if (!parsedStates || !parsedCSFSSamples) {
        cerr << "ERROR. Parsed " << int(currentType) << " before parsing states and number of CSFS samples." << endl;
        exit(1);
      }
      int undistinguishedIndex = StringUtils::stoi(splitString[1]);
      vector < vector <float> > thisCSFS = vector < vector <float> > ();
      for (int k = 0; k < 2; k++) {
        getline(br, line);
        vector <string> inSplitString; istringstream iss(line); string buf; while (iss >> buf) inSplitString.push_back(buf);
        if (inSplitString.size() != states) {
          cerr << "ERROR. Parsed " << inSplitString.size() + " FoldedCSFS entries for " << states << " states: " << line << endl;
          exit(1);
        }
        vector <float> CSFSline = vector <float> (states);
        thisCSFS.push_back(CSFSline);
        for (unsigned int i = 0; i < states; i++) {
          thisCSFS[k][i] = StringUtils::stof(inSplitString[i]);
        }
      }
      foldedCSFSmap[undistinguishedIndex] = thisCSFS;
      // cout << "Parsed FoldedCSFS " << undistinguishedIndex << " " << foldedCSFSmap[undistinguishedIndex].size() << " " << foldedCSFSmap[undistinguishedIndex][0].size() << endl;
      continue;
    } else if (firstToken == string("ascertainedcsfs")) {
      currentType = DataType::AscertainedCSFS;
      if (!parsedStates || !parsedCSFSSamples) {
        cerr << "ERROR. Parsed " << int(currentType) << " before parsing states and number of CSFS samples." << endl;
        exit(1);
      }
      int undistinguishedIndex = StringUtils::stoi(splitString[1]);
      vector < vector <float> > thisCSFS = vector < vector <float> > ();
      for (int k = 0; k < 3; k++) {
        getline(br, line);
        vector <string> inSplitString; istringstream iss(line); string buf; while (iss >> buf) inSplitString.push_back(buf);
        if (inSplitString.size() != states) {
          cerr << "ERROR. Parsed " << inSplitString.size() + " AscertainedCSFS entries for " << states << " states: " << line << endl;
          exit(1);
        }
        vector <float> CSFSline = vector <float> (states);
        thisCSFS.push_back(CSFSline);
        for (unsigned int i = 0; i < states; i++) {
          thisCSFS[k][i] = StringUtils::stof(inSplitString[i]);
        }
      }
      ascertainedCSFSmap[undistinguishedIndex] = thisCSFS;
      // cout << "Parsed AscertainedCSFS " << undistinguishedIndex << " " << ascertainedCSFSmap[undistinguishedIndex].size() << " " << ascertainedCSFSmap[undistinguishedIndex][0].size() << endl;
      continue;
    } else if (firstToken == string("foldedascertainedcsfs")) {
      currentType = DataType::FoldedAscertainedCSFS;
      if (!parsedStates || !parsedCSFSSamples) {
        cerr << "ERROR. Parsed " << int(currentType) << " before parsing states and number of CSFS samples." << endl;
        exit(1);
      }
      int undistinguishedIndex = StringUtils::stoi(splitString[1]);
      vector < vector <float> > thisCSFS = vector < vector <float> > ();
      for (int k = 0; k < 2; k++) {
        getline(br, line);
        vector <string> inSplitString; istringstream iss(line); string buf; while (iss >> buf) inSplitString.push_back(buf);
        if (inSplitString.size() != states) {
          cerr << "ERROR. Parsed " << inSplitString.size() + " FoldedAscertainedCSFS entries for " << states << " states: " << line << endl;
          exit(1);
        }
        vector <float> CSFSline = vector <float> (states);
        thisCSFS.push_back(CSFSline);
        for (unsigned int i = 0; i < states; i++) {
          thisCSFS[k][i] = StringUtils::stof(inSplitString[i]);
        }
      }
      foldedAscertainedCSFSmap[undistinguishedIndex] = thisCSFS;
      // cout << "Parsed FoldedAscertainedCSFS " << undistinguishedIndex << " " << foldedAscertainedCSFSmap[undistinguishedIndex].size() << " " << foldedAscertainedCSFSmap[undistinguishedIndex][0].size() << endl;
      continue;
    } else if (firstToken == string("homozygousemissions")) {
      currentType = DataType::HomozygousEmissions;
    } else if (firstToken == string("initialstateprob")) {
      currentType = DataType::initialStateProb;
    } else if (firstToken == string("columnratios")) {
      currentType = DataType::ColumnRatios;
    } else if (firstToken == string("rowratios")) {
      currentType = DataType::RowRatios;
    } else if (firstToken == string("uvectors")) {
      currentType = DataType::Uvectors;
    } else if (firstToken == string("bvectors")) {
      currentType = DataType::Bvectors;
    } else if (firstToken == string("dvectors")) {
      currentType = DataType::Dvectors;
    } else {
      // reading a content line
      if (currentType == DataType::ColumnRatios) {
        columnRatios = vector <float> (states, 0.0);
        for (unsigned int i = 0; i < splitString.size(); i++) {
          columnRatios[i] = StringUtils::stof(splitString[i]);
        }
        // cout << "Parsed ColumnRatios" << endl;
      } else if (currentType == DataType::initialStateProb) {
        initialStateProb = vector <float> (states, 0.0);
        for (unsigned int i = 0; i < splitString.size(); i++) {
          initialStateProb[i] = StringUtils::stof(splitString[i]);
        }
        // cout << "Parsed initialStateProb" << endl;
      } else if (currentType == DataType::RowRatios) {
        vector <float> thisRow(states, 0.0);
        for (unsigned int i = 1; i < splitString.size(); i++) {
          thisRow[i-1] = StringUtils::stof(splitString[i]);
        }
        float index = StringUtils::stof(splitString[0]);
        rowRatioVectors[index] = thisRow;
        // cout << "Parsed RowRatios" << endl;
      } else if (currentType == DataType::Uvectors) {
        vector <float> thisRow(states, 0.0);
        for (unsigned int i = 1; i < splitString.size(); i++) {
          thisRow[i-1] = StringUtils::stof(splitString[i]);
        }
        float index = StringUtils::stof(splitString[0]);
        Uvectors[index] = thisRow;
        // cout << "Parsed Uvectors" << endl;
      } else if (currentType == DataType::Bvectors) {
        vector <float> thisRow(states, 0.0);
        for (unsigned int i = 1; i < splitString.size(); i++) {
          thisRow[i-1] = StringUtils::stof(splitString[i]);
        }
        float index = StringUtils::stof(splitString[0]);
        Bvectors[index] = thisRow;
        // cout << "Parsed Bvectors" << endl;
      } else if (currentType == DataType::Dvectors) {
        vector <float> thisRow(states, 0.0);
        for (unsigned int i = 1; i < splitString.size(); i++) {
          thisRow[i-1] = StringUtils::stof(splitString[i]);
        }
        float index = StringUtils::stof(splitString[0]);
        Dvectors[index] = thisRow;
        // cout << "Parsed Dvectors" << endl;
      } else if (currentType == DataType::HomozygousEmissions) {
        vector <float> thisRow(states, 0.0);
        for (unsigned int i = 1; i < splitString.size(); i++) {
          thisRow[i-1] = StringUtils::stof(splitString[i]);
        }
        int index = StringUtils::stoi(splitString[0]);
        homozygousEmissionMap[index] = thisRow;
        // cout << "Parsed HomozygousEmissions" << endl;
      }
    }
  }
}

File: TOOLS/PREPARE_DECODING/src/ASMCprepareDecoding/DecodingQuantities.java
//    This file is part of ASMC, developed by Pier Francesco Palamara.
//
//    ASMC is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    ASMC is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with ASMC.  If not, see <https://www.gnu.org/licenses/>.


package ASMCprepareDecoding;

import java.io.BufferedWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.TreeMap;
import static ASMCprepareDecoding.CSFS.loadFromFile;
import org.jblas.DoubleMatrix;

public class DecodingQuantities {

    // these are the quantities loaded from file, which are required for the linear-time decoding
    Transition.TransitionType transitionType;
    DoubleMatrix initialStateProb;
    double[] timeVector;
    double[] sizeVector;
    double[] discretization;
    double[] expectedTimes;
    ArrayList<Double> geneticDistancesList = new ArrayList<Double>();
    ArrayList<Integer> physDistancesList = new ArrayList<Integer>();
    ArrayList<DoubleMatrix> DvectorMap = new ArrayList<DoubleMatrix>();
    ArrayList<DoubleMatrix> BvectorMap = new ArrayList<DoubleMatrix>();
    ArrayList<DoubleMatrix> UvectorMap = new ArrayList<DoubleMatrix>();
    ArrayList<DoubleMatrix> rowRatiosVectorMap = new ArrayList<DoubleMatrix>();
    ArrayList<DoubleMatrix> HomozygousEmissionMap = new ArrayList<DoubleMatrix>();
    DoubleMatrix columnRatios;

    private int states, CSFSSamples;
    double mu;

    // sequence emission
    TreeMap<Double, CSFSEntry> CSFSmap;
    TreeMap<Double, CSFSEntry> foldedCSFSmap;
    private double[][] classicEmissionTable;

    // ascertained emission
    TreeMap<Double, CSFSEntry> ascertainedCSFSmap;
    TreeMap<Double, CSFSEntry> foldedAscertainedCSFSap;
    private double[][] compressedEmissionTable;

    // for a number between 10^(n) and 10^(n+1), proceed in steps of size max(1, 10^(n-precision))
    final static int precision = 2;
    final static double minGenetic = 1e-10;

    // start/end for physical and genetic distances (in Morgans)
    private static final double startGen = 1e-10;
    private static final double maxGen = 0.3; // 30 centiMorgans
    private static final int startPhys = 1;
    private static final int maxPhys = 100000000; // 100 Mb

    // for each genetic distance, compute decoding quantities
    public DecodingQuantities(CSFS csfs, Transition transition, double mu) throws IOException {
        this.timeVector = transition.timeVector;
        this.mu = mu;
        this.sizeVector = transition.sizeVector;
        this.CSFSSamples = csfs.samples;
        this.discretization = transition.discretization;
        this.expectedTimes = transition.expectedTimes;
        this.states = transition.discretization.length - 1;
        int lastPercentage = -1;
        columnRatios = transition.columnRatios;

        // compute transition quantities
        // putting 0 in. This is not really used, depending on the value of smallNumber in the decoding side.
        computeTransitionQuantitiesForOnePosition(0.0, transition);
        geneticDistancesList.add(0.0);
        double genDist = startGen;
        while (genDist < maxGen) {
            geneticDistancesList.add(genDist);
            genDist = DecodingQuantities.nextGen(genDist);
        }
        // start at 1, 0 is done already
        for (int i = 1; i < geneticDistancesList.size(); i++) {
            genDist = geneticDistancesList.get(i);
            // compute emission and transition quantities in linear time
            computeTransitionQuantitiesForOnePosition(genDist, transition);
            int percentage = (int) Math.round(100 * i / (float) geneticDistancesList.size());
            if (percentage != lastPercentage) {
                Utils.printf("\rGenetic distances progress: " + percentage + "%");
            }
            lastPercentage = percentage;
        }
        Utils.print("");

        // compute homozygous emissions
        int phys = startPhys;
        while (phys < maxPhys) {
            this.physDistancesList.add(phys);
            phys = DecodingQuantities.nextPhys(phys);
        }
        for (int i = 0; i < physDistancesList.size(); i++) {
            HomozygousEmissionMap.add(new DoubleMatrix(CSFS.computeClassicEmission(this.expectedTimes, physDistancesList.get(i) * this.mu)));
            int percentage = (int) Math.round(100 * i / (float) physDistancesList.size());
            if (percentage != lastPercentage) {
                Utils.printf("\rPhysical distances progress: " + percentage + "%");
            }
            lastPercentage = percentage;
        }
        Utils.print("");

        // get initial state distribution from coalescent distribution
        double lastProb = 0;
        DoubleMatrix cumProb = new DoubleMatrix(this.states);
        initialStateProb = new DoubleMatrix(this.states);
        for (int i = 0; i < this.states; i++) {
            double timeT = transition.discretization[i + 1];
            cumProb.put(i, transition.cumulativeCoalesceFromStoT(0, timeT));
            initialStateProb.put(i, cumProb.get(i) - lastProb);
            lastProb = cumProb.get(i);
        }

        this.transitionType = transition.type;

        this.CSFSmap = csfs.CSFS;
        this.foldedCSFSmap = csfs.foldedCSFS;
        this.classicEmissionTable = CSFS.computeClassicEmission(this.expectedTimes, mu);

        this.ascertainedCSFSmap = csfs.ascertainedCSFS;
        this.foldedAscertainedCSFSap = csfs.foldedAscertainedCSFS;
        this.compressedEmissionTable = csfs.compressedAscertainedEmissionTable;
    }

    public static int nextPhys(int phys) {
        if (phys < 0) {
            Utils.exit("Int overflow: " + phys);
        }
        int log10 = (int) Math.max(0, Math.floor(Math.log10(phys)) - precision);
        int factor = (int) Math.pow(10, log10);
        int next = Math.round(phys / (float) factor + 1) * factor;
        return next;
    }

    public static int roundPhysical(int phys) {
        if (phys < 0) {
            Utils.exit("Int overflow: " + phys);
        }
        int log10 = (int) Math.max(0, Math.floor(Math.log10(phys)) - precision);
        int factor = (int) Math.pow(10, log10);
        int rounded = Math.round(phys / (float) factor) * factor;
        return rounded;
    }

    public static double nextGen(double gen) {
        double gene1e10 = gen * 1e10;
        int log10 = (int) Math.max(0, Math.floor(Math.log10(gene1e10)) - precision);
        double factor = Math.pow(10, log10);
        double next = (Math.round(gene1e10 / factor) + 1) * factor;
        return next / 1e10;
    }

    public static double roundMorgans(double gen) {
        double gene1e10 = gen * 1e10;
        int log10 = (int) Math.max(0, Math.floor(Math.log10(gene1e10)) - precision);
        double factor = Math.pow(10, log10);
        double rounded = Math.round(gene1e10 / factor) * factor;
        return Math.max(minGenetic, rounded / 1e10);
    }

    // compute quantities for a give genetic distance
    private void computeTransitionQuantitiesForOnePosition(double genDist, Transition transition) {
        ArrayList<DoubleMatrix> res = transition.getLinearTimeDecodingQuantitiesAndMatrixGivenDistance(genDist);
        // index 0 is transition matrix
        DoubleMatrix D = res.get(0);
        DoubleMatrix B = res.get(1);
        DoubleMatrix U = res.get(2);
        DoubleMatrix RR = res.get(3);
        DvectorMap.add(D);
        BvectorMap.add(B);
        UvectorMap.add(U);
        rowRatiosVectorMap.add(RR);
    }

    // save all compute quantities to file
    public void saveDecodingQuantities(String outputFileRoot) throws IOException {
        BufferedWriter outBuff = Utils.openGzipFileForWriting(outputFileRoot + ".decodingQuantities.gz");
        // write model parameters
        outBuff.write("TransitionType\n" + this.transitionType + "\n\n");
        outBuff.write("States\n" + this.states + "\n\n");
        outBuff.write("CSFSSamples\n" + this.CSFSSamples + "\n\n");
        outBuff.write("TimeVector\n");
        outBuff.write(Utils.doubleMatrixToString(timeVector));
        outBuff.write("\n");
        outBuff.write("SizeVector\n");
        outBuff.write(Utils.doubleMatrixToString(sizeVector));
        outBuff.write("\n");
        outBuff.write("Discretization\n");
        outBuff.write(Utils.doubleMatrixToString(discretization));
        outBuff.write("\n");
        outBuff.write("ExpectedTimes\n");
        outBuff.write(Utils.doubleMatrixToString(expectedTimes));
        // write sequence Emissions
        outBuff.write("\n");
        for (int undistinguished = 0; undistinguished < this.CSFSmap.firstEntry().getValue().CSFS[0].length; undistinguished++) {
            outBuff.write("CSFS\t" + undistinguished + "\n");
            for (int distinguished = 0; distinguished < 3; distinguished++) {
                for (double from : this.CSFSmap.keySet()) {
                    outBuff.write(this.CSFSmap.get(from).CSFS[distinguished][undistinguished] + "\t");
                }
                outBuff.write("\n");
            }
        }
        outBuff.write("\n");
        for (int undistinguished = 0; undistinguished < this.foldedCSFSmap.firstEntry().getValue().CSFS[0].length; undistinguished++) {
            outBuff.write("FoldedCSFS\t" + undistinguished + "\n");
            for (int distinguished = 0; distinguished < 2; distinguished++) {
                for (double from : this.foldedCSFSmap.keySet()) {
                    outBuff.write(this.foldedCSFSmap.get(from).CSFS[distinguished][undistinguished] + "\t");
                }
                outBuff.write("\n");
            }
        }
        outBuff.write("\n");
        outBuff.write("ClassicEmission\n");
        outBuff.write(Utils.doubleMatrixToString(classicEmissionTable));
        // write ascertained Emissions
        outBuff.write("\n");
        for (int undistinguished = 0; undistinguished < this.ascertainedCSFSmap.firstEntry().getValue().CSFS[0].length; undistinguished++) {
            outBuff.write("AscertainedCSFS\t" + undistinguished + "\n");
            for (int distinguished = 0; distinguished < 3; distinguished++) {
                for (double from : this.ascertainedCSFSmap.keySet()) {
                    outBuff.write(this.ascertainedCSFSmap.get(from).CSFS[distinguished][undistinguished] + "\t");
                }
                outBuff.write("\n");
            }
        }
        outBuff.write("\n");
        for (int undistinguished = 0; undistinguished < this.foldedAscertainedCSFSap.firstEntry().getValue().CSFS[0].length; undistinguished++) {
            outBuff.write("FoldedAscertainedCSFS\t" + undistinguished + "\n");
            for (int distinguished = 0; distinguished < 2; distinguished++) {
                for (double from : this.foldedAscertainedCSFSap.keySet()) {
                    outBuff.write(this.foldedAscertainedCSFSap.get(from).CSFS[distinguished][undistinguished] + "\t");
                }
                outBuff.write("\n");
            }
        }
        outBuff.write("\n");
        outBuff.write("CompressedAscertainedEmission\n");
        outBuff.write(Utils.doubleMatrixToString(compressedEmissionTable));
        // write initial state distribution
        outBuff.write("\n");
        outBuff.write("initialStateProb\n");
        outBuff.write(Utils.doubleMatrixToString(initialStateProb.transpose()));
        // write column ratios
        outBuff.write("\n");
        outBuff.write("ColumnRatios\n");
        outBuff.write(Utils.doubleMatrixToString(columnRatios.transpose()));
        // write row ratios
        outBuff.write("\n");
        outBuff.write("RowRatios\n");
        for (int i = 0; i < rowRatiosVectorMap.size(); i++) {
            DoubleMatrix thisRR = rowRatiosVectorMap.get(i);
            outBuff.write(geneticDistancesList.get(i) + "\t" + Utils.doubleMatrixToString(thisRR.transpose()));
        }
        // write U vectors
        outBuff.write("\n");
        outBuff.write("Uvectors\n");
        for (int i = 0; i < UvectorMap.size(); i++) {
            DoubleMatrix thisU = UvectorMap.get(i);
            outBuff.write(geneticDistancesList.get(i) + "\t" + Utils.doubleMatrixToString(thisU.transpose()));
        }
        // write B vectors
        outBuff.write("\n");
        outBuff.write("Bvectors\n");
        for (int i = 0; i < BvectorMap.size(); i++) {
            DoubleMatrix thisB = BvectorMap.get(i);
            outBuff.write(geneticDistancesList.get(i) + "\t" + Utils.doubleMatrixToString(thisB.transpose()));
        }
        // write D vectors
        outBuff.write("\n");
        outBuff.write("Dvectors\n");
        for (int i = 0; i < DvectorMap.size(); i++) {
            DoubleMatrix thisD = DvectorMap.get(i);
            outBuff.write(geneticDistancesList.get(i) + "\t" + Utils.doubleMatrixToString(thisD.transpose()));
        }
        // write homozygous emissions
        outBuff.write("\n");
        outBuff.write("HomozygousEmissions\n");
        for (int i = 0; i < physDistancesList.size(); i++) {
            int phys = physDistancesList.get(i);
            outBuff.write(phys + "\t" + Utils.doubleMatrixToString(HomozygousEmissionMap.get(i).getRow(0)));
        }
        outBuff.close();
    }

//    // unit test
//    public static void unitTest() throws IOException {
//        Data data = new Data("CHR22_100_1/out.1234.CEU.full.n100.chr22.len1.array");
//        double[] discretization = {0.0, 1118.2, 1472.2, 1849.7, 2497.0, 3963.8, 9120.8, 15832.9, 24139.9, 34891.6, Double.POSITIVE_INFINITY};
//        Transition transition = new Transition(Transition.EUtime_array, Transition.EUsize_array, discretization, Transition.TransitionType.CSC);
//        CSFS csfs = loadFromFile("CHR22_100_1/out.1234.CEU.full.n100.chr22.len1.array.csfs");
//        csfs.fixAscertainment(data, 100, transition);
//        double mutRate = 1.65E-8;
//        DecodingQuantities decodingQuantities = new DecodingQuantities(csfs, transition, mutRate);
//        decodingQuantities.saveDecodingQuantities("CHR22_100_1/out.1234.CEU.full.n100.chr22.len1.array.temp");
//    }

}
